Add flow control for websocket reader

mehaase opened the pull request:
The patch for flow control is 3 lines. I updated the documentation in a separate commit, in case you want to cherry pick the patch by itself. I tried really hard to think of a way to write an automated test for this, but couldn't come up with anything that works. (You weren't kidding when you said that unit testing `asyncio` code is hard.)

aaugustin commented:
Thanks!

I think 1024 is too high. Each message can be up to 1MB utf-8 encoded which, due to Python's string handling, can take 4MB in memory (in pathological circumstances, but that's what we must consider for security). So the queue for each worker can theoretical use up to 4 * max_queue MB. That's why I think the limit should be something like 32, which means the effective limit is 128MB of memory per worker.

That's still a lot of memory. To get more security a developer should really set max_size to what they really need. But I don't want to make that limit low by default because that's likely to cause unexpected problems and give the impression that the library doesn't work well.

aaugustin commented:
I can merge your proposal and make the small adjustments myself.

aaugustin commented:
By the way, did you get the opportunity to redo your load test with the patch and confirm that it solves the problem of excessive memory use?

mehaase commented:
I did run the test again. I modified the client and the server to print out a sequence number (`print i; i += 1`) after every send or receive. Then I set max_queue to about 100. Without the patch, the server sends packets as fast as it can, and the client's resident set size (as seen in top) climbs steadily by a few megabytes per second.

With the patch, the server sends a flurry of about ~150 packets and then pauses. After the client has processed maybe 10-20 messages, the server sends another little flurry. I let this run over night (33,188 packets now processed) and the client's resident set size has remained constant. It clearly solves the resource consumption problem.

This morning, I wondered if any packets would be dropped. So I modified the server to include an ID number in each message and the client to print out the ID of each message and check that no IDs are skipped. (I'll edit my original post to include these updated versions, in case you want to try it yourself.) After letting this run for about 10 minutes, the client has received all of the messages and resident set size has remained constant. I think this is good evidence that flow control is working.

aaugustin commented:
Great work! Thanks for taking the time to experiment.

I'll think a bit about tests and merge this soon.

If you don't see any movement in two weeks, poke me ;-)

aaugustin commented:
I merged your changes, extended the documentation a bit, and released version 3.1. Thanks for your contribution!

aaugustin commented:
That's where I realize I don't have a good way to credit contributors... Oops :-/

mehaase commented:
No credit necessary! It helped me solve a problem in my own project, and I'm happy to contribute to a good open source project.
